File: src/Warehouse.Core/Markets/Concrete/Okx/OkxSynchronizationWorker.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Warehouse.Core.Markets.Concrete.Okx.Constants;
using Warehouse.Core.Markets.Concrete.Okx.Messages.Http;
using Warehouse.Core.Markets.Concrete.Okx.Services;
using Warehouse.Core.Markets.Domain;
using Warehouse.Core.Shared;
using Warehouse.Core.Shared.Domain;
using Warehouse.Core.Shared.Services;
using static Warehouse.Core.Shared.Domain.Instrument;

namespace Warehouse.Core.Markets.Concrete.Okx;

public class OkxSynchronizationWorker(IServiceScopeFactory scopeFactory, ILogger<OkxSynchronizationWorker> logger) : BackgroundService
{
    private const string Timeframe = CandlestickTimeframes.OneMinute;
    private const int BatchSize = 100;
    private static readonly Instrument[] Symbols = [OKB, BTC, SOL, ETH, DOGE, XRP, BCH, LTC];
    private static readonly DateTime SyncStartDate = DateTime.Today;
    private readonly PeriodicTimer periodicTimer = new(TimeSpan.FromMinutes(1));

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        List<Candlestick> candlesticks = [];
        while (await periodicTimer.WaitForNextTickAsync(stoppingToken))
        {
            await using AsyncServiceScope scope = scopeFactory.CreateAsyncScope();
            ICandlestickService candlestickService = scope.ServiceProvider.GetRequiredService<ICandlestickService>();
            OkxHttpService okxHttpService = scope.ServiceProvider.GetRequiredService<OkxHttpService>();

            foreach (Instrument symbol in Symbols)
            {
                var pairSymbol = new Pair(symbol, USDT);
                Candlestick? latestCandle = await candlestickService.GetLatestCandlestickAsync(
                    pairSymbol,
                    MarketType.Okx,
                    Timeframe,
                    stoppingToken);
                DateTime startFrom = latestCandle?.Timestamp.AddMinutes(-2) ?? SyncStartDate;

                await foreach (Candlestick candlestick in GetCandlesticksAsync(pairSymbol, startFrom, okxHttpService)
                                   .WithCancellation(stoppingToken))
                {
                    candlesticks.Add(candlestick);
                }
            }

            int saved = await candlestickService.SaveCandlesticksAsync(candlesticks, stoppingToken);

            logger.LogInformation(
                "Saved batch of {Count} candlesticks. Latest: {LatestTimestamp}.",
                saved,
                candlesticks.Max(x => x.Timestamp));

            candlesticks.Clear();
        }
    }

    private async IAsyncEnumerable<Candlestick> GetCandlesticksAsync(string symbol, DateTime fromDate, OkxHttpService okxHttpService)
    {
        logger.LogInformation("Starting candlestick sync for {Symbol} from {FromDate}", symbol, fromDate);

        var before = new DateTimeOffset(fromDate).ToUnixTimeMilliseconds().ToString();

        Result<OkxCandlestick[]> result = await okxHttpService.GetCandlesticksAsync(symbol, Timeframe, before: before, limit: BatchSize);

        if (!result.IsSuccess)
        {
            logger.LogError("Failed to fetch candlesticks: {Error}", result.Error);
            yield break;
        }

        OkxCandlestick[]? okxCandles = result.Value;

        if (okxCandles is null || okxCandles.Length == 0)
        {
            logger.LogInformation("No more candlesticks to fetch");
            yield break;
        }

        logger.LogDebug("Fetched {Count} candlesticks", okxCandles.Length);

        foreach (OkxCandlestick okxCandle in okxCandles)
        {
            yield return new Candlestick
            {
                Symbol = symbol,
                MarketType = MarketType.Okx,
                Timestamp = okxCandle.Timestamp,
                Open = okxCandle.Open,
                High = okxCandle.High,
                Low = okxCandle.Low,
                Close = okxCandle.Close,
                Volume = okxCandle.Volume,
                VolumeQuote = okxCandle.VolumeQuoteCurrency,
                IsCompleted = okxCandle.IsCompleted,
                Timeframe = Timeframe
            };
        }
    }
}

File: src/Warehouse.Core/Markets/Concrete/Okx/OkxSynchronizationWorker.fs
namespace Warehouse.Core.Markets.Concrete.Okx

open System
open System.Threading
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Warehouse.Core.Functional.Shared.Domain
open Warehouse.Core.Functional.Markets.Domain
open Warehouse.Core.Functional.Markets.Contracts
open Warehouse.Core.Markets.Concrete.Okx.Services
open FSharp.Control

type OkxSynchronizationWorker(scopeFactory: IServiceScopeFactory, logger: ILogger<OkxSynchronizationWorker>) =
    inherit BackgroundService()

    let timeframe = "1m" // CandlestickTimeframes.OneMinute
    let batchSize = 100

    let symbols =
        [|
            Instrument.OKB
            Instrument.BTC
            Instrument.SOL
            Instrument.ETH
            Instrument.DOGE
            Instrument.XRP
            Instrument.BCH
            Instrument.LTC
        |]

    let syncStartDate = DateTime.Today
    let periodicTimer = new PeriodicTimer(TimeSpan.FromMinutes(1.0))

    let getCandlesticksAsync (symbol: string) (fromDate: DateTime) (okxHttpService: OkxHttpService) =
        taskSeq {
            logger.LogInformation("Starting candlestick sync for {Symbol} from {FromDate}", symbol, fromDate)

            let before = DateTimeOffset(fromDate).ToUnixTimeMilliseconds().ToString()

            let! result =
                okxHttpService.GetCandlesticksAsync(symbol, bar = timeframe, before = before, limit = batchSize)
                |> Async.AwaitTask

            match result with
            | Ok result ->
                let okxCandles = result

                if okxCandles.Length = 0 then
                    logger.LogInformation("No more candlesticks to fetch")
                    ()
                else
                    logger.LogDebug("Fetched {Count} candlesticks", okxCandles.Length)

                    for okxCandle in okxCandles do
                        yield
                            {
                                Id = 0
                                Symbol = symbol
                                MarketType = int MarketType.Okx
                                Timestamp = okxCandle.Timestamp
                                Open = okxCandle.Open
                                High = okxCandle.High
                                Low = okxCandle.Low
                                Close = okxCandle.Close
                                Volume = okxCandle.Volume
                                VolumeQuote = okxCandle.VolumeQuoteCurrency
                                IsCompleted = okxCandle.IsCompleted
                                Timeframe = timeframe
                            }
            | Error error -> logger.LogError("Failed to fetch candlesticks: {Error}", error)
        }

    override this.ExecuteAsync(stoppingToken: CancellationToken) =
        task {
            let candlesticks = ResizeArray<Candlestick>()

            while not stoppingToken.IsCancellationRequested do
                let! tick = periodicTimer.WaitForNextTickAsync(stoppingToken)

                if tick then
                    use scope = scopeFactory.CreateAsyncScope()
                    let candlestickService = scope.ServiceProvider.GetRequiredService<ICandlestickService>()
                    let okxHttpService = scope.ServiceProvider.GetRequiredService<OkxHttpService>()

                    for symbol in symbols do
                        let pairSymbol = { Left = symbol; Right = Instrument.USDT }.ToString()

                        let! latestCandle =
                            candlestickService.GetLatestCandlestickAsync(
                                pairSymbol,
                                MarketType.Okx,
                                timeframe,
                                stoppingToken
                            )

                        let startFrom =
                            match latestCandle with
                            | Some c -> c.Timestamp.AddMinutes(-2.0)
                            | None -> syncStartDate

                        // Converting AsyncSeq to Task/Loop
                        // System.Threading.Tasks.Extensions or similar providing AsyncEnumerable?
                        // For simplicity using loop if GetCandlesticksAsync was enumerable
                        // But I defined it as asyncSeq. F# Control.AsyncSeq is good but might need library.
                        // Let's rewrite `getCandlesticksAsync` to return Task<seq<Candlestick>> or similar to avoid extra dependencies if AsyncSeq not present. Use simple recursion or loop inside task.

                        // Re-implementing inner logic inline or via simple helper strictly Task based.

                        let rec fetchBatch fromDate =
                            task {
                                let before = DateTimeOffset(fromDate).ToUnixTimeMilliseconds().ToString()

                                let! result =
                                    okxHttpService.GetCandlesticksAsync(
                                        pairSymbol,
                                        bar = timeframe,
                                        before = before,
                                        limit = batchSize
                                    )

                                match result with
                                | Error error ->
                                    // logger.LogError("Failed to fetch candlesticks: {Error}", error.Message)
                                    return [||]
                                | Ok okxCandles -> return okxCandles
                            }

                        // Just one batch for now per cycle per symbol?
                        // The original C# code does `await foreach` so it might fetch multiple batches?
                        // "GetCandlesticksAsync" in C# yields items from *one* result of "GetCandlesticksAsync".
                        // Ah, the C# GetCandlesticksAsync method calls okxHttpService.GetCandlesticksAsync ONCE.
                        // So it is NOT a recursive historical fetcher in the loop. It just fetches one batch.
                        // Wait, `await foreach (Candlestick candlestick in GetCandlesticksAsync(...))` inside the loop.
                        // And `GetCandlesticksAsync` calls `okxHttpService.GetCandlesticksAsync` ONCE.
                        // So it is fundamentally fetching ONE batch of 100 candles every minute?
                        // Yes, looks like a synchronization of RECENT candles, not full history.

                        let! batch = fetchBatch startFrom

                        if batch.Length > 0 then
                            logger.LogDebug("Fetched {Count} candlesticks for {Symbol}", batch.Length, pairSymbol)

                            for okxCandle in batch do
                                candlesticks.Add(
                                    {
                                        Id = 0
                                        Symbol = pairSymbol
                                        MarketType = int MarketType.Okx
                                        Timestamp = okxCandle.Timestamp
                                        Open = okxCandle.Open
                                        High = okxCandle.High
                                        Low = okxCandle.Low
                                        Close = okxCandle.Close
                                        Volume = okxCandle.Volume
                                        VolumeQuote = okxCandle.VolumeQuoteCurrency
                                        IsCompleted = okxCandle.IsCompleted
                                        Timeframe = timeframe
                                    }
                                )

                    if candlesticks.Count > 0 then
                        let! saved = candlestickService.SaveCandlesticksAsync(candlesticks, stoppingToken)

                        logger.LogInformation(
                            "Saved batch of {Count} candlesticks. Latest: {LatestTimestamp}.",
                            saved,
                            candlesticks |> Seq.map _.Timestamp |> Seq.max
                        )

                        candlesticks.Clear()
        }
